pkg/executor/infoschema_reader.go

```diff
@@ -402,36 +399,38 @@ func (e *memtableRetriever) setDataFromSchemata(ctx sessionctx.Context, schemas
 		rows = append(rows, record)
 	}
 	e.rows = rows
+	return nil
 }
 
-func (e *memtableRetriever) setDataForStatistics(ctx context.Context, sctx sessionctx.Context, schemas []model.CIStr) error {
+func (e *memtableRetriever) setDataForStatistics(ctx context.Context, sctx sessionctx.Context) error {
 	checker := privilege.GetPrivilegeManager(sctx)
-	extractor, ok := e.extractor.(*plannercore.InfoSchemaTablesExtractor)
-	if ok && extractor.SkipRequest {
+	ex, ok := e.extractor.(*plannercore.InfoSchemaStatisticsExtractor)
+	if !ok {
+		return errors.Errorf("wrong extractor type: %T, expected InfoSchemaStatisticsExtractor", e.extractor)
+	}
+	if ex.SkipRequest {
 		return nil
 	}
+	schemas := ex.ListSchemas(e.is)
 	for _, schema := range schemas {
-		if ok && extractor.Filter("table_schema", schema.L) {
-			continue
-		}
-		tables, err := e.is.SchemaTableInfos(ctx, schema)
+		tables, err := ex.ListTables(ctx, e.is, schema)
```

Comment:
Why ListTables?

Reply:
here `List`... does not mean full scan tables. It will filter with `ex`'s predicates. Maybe rename it?

Reply:
Do you have any suggestions for naming?

It is actually "list part of tables if predicate exists, otherwise list all tables".

Reply:
Do you mean `FilteredSchema` and `FilteredTables`?